sdk/src/activity_context.rs

```diff
@@ -0,0 +1,202 @@
+    pub task_queue: String,
+    pub heartbeat_timeout: Option<Duration>,
+    /// Time of activity scheduled by a workflow
+    pub scheduled_time: Option<Timestamp>,
+    /// Time of activity start
+    pub started_time: Option<Timestamp>,
+    /// Time of activity timeout
+    pub deadline: Option<Timestamp>,
+    /// Attempt starts from 1, and increased by 1 for every retry if retry policy is specified.
+    pub attempt: u32,
+
+    /// #NOTE Not in sdk-go
+    pub current_attempt_scheduled_time: Option<Timestamp>,
+    pub retry_policy: Option<RetryPolicy>,
+    pub is_local: bool,
+}
+
+impl ActContext {
+    /// Construct new Activity Context
+    pub fn new(
+        worker: Arc<dyn Worker>,
+        task_queue: String,
+        task_token: Vec<u8>,
+        task: activity_task::Start,
+    ) -> Self {
+        let activity_task::Start {
+            workflow_namespace,
+            workflow_type,
+            workflow_execution,
+            activity_id,
+            activity_type,
+            header_fields,
+            input,
+            heartbeat_details,
+            scheduled_time,
+            current_attempt_scheduled_time,
+            started_time,
+            attempt,
+            schedule_to_close_timeout,
+            start_to_close_timeout,
+            heartbeat_timeout,
+            retry_policy,
+            is_local,
+        } = task;
+        let deadline = calculate_deadline(
+            scheduled_time.as_ref(),
+            started_time.as_ref(),
+            start_to_close_timeout.as_ref(),
+            schedule_to_close_timeout.as_ref(),
+        );
+
+        ActContext {
+            worker,
+            input,
+            heartbeat_details,
+            header_fields,
+            info: ActivityInfo {
+                task_token,
+                task_queue,
+                workflow_type,
+                workflow_namespace,
+                workflow_execution,
+                activity_id,
+                activity_type,
+                heartbeat_timeout,
+                scheduled_time,
+                started_time,
+                deadline,
+                attempt,
+
+                current_attempt_scheduled_time,
+                retry_policy,
+                is_local,
+            },
+        }
+    }
+
+    /// Retrieve the rest of the parameters
+    /// #NOTE the first parameter is popped for convenience, the rest is still stored in the
+    /// activity state.  There are no variadic parameters in Rust.
+    pub fn inputs(&mut self) -> &mut [Payload] {
+        &mut self.input
+    }
+
+    /// Extract heartbeat details from last failed attempt. This is used in combination with retry policy.
+    pub fn get_heartbeat_details(&self) -> &[Payload] {
+        &self.heartbeat_details
+    }
+
+    /// Helper function to check if details exist from previous execution of this activity
+    pub fn has_heartbeat_details(&self) -> bool {
+        !self.heartbeat_details.is_empty()
+    }
```

Comment:
Given  `get_heartbeat_details` returns a ref, I don't know that we need this extra function

Reply:
Reasonable.  I was trying to keep the API's in line but yes, with the reference, callers can check themselves without the helper.  Fixed